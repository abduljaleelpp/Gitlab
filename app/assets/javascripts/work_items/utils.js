import workItemQuery from './graphql/work_item.query.graphql';
import workItemByIidQuery from './graphql/work_item_by_iid.query.graphql';

export function getWorkItemQuery(isFetchedByIid) {
  return isFetchedByIid ? workItemByIidQuery : workItemQuery;
}
