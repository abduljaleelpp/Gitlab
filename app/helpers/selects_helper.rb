# frozen_string_literal: true

module SelectsHelper
  def groups_select_tag(id, opts = {})
    classes = Array.wrap(opts[:class])
    classes << 'ajax-groups-select'

    # EE requires this line to be present, but there is no easy way of injecting
    # this into EE without causing merge conflicts. Given this line is very
    # simple and not really EE specific on its own, we just include it in CE.
    classes << 'multiselect' if opts[:multiple]

    opts[:class] = classes.join(' ')

    select2_tag(id, opts)
  end

  def project_select_tag(id, opts = {})
    opts[:class] = [*opts[:class], 'ajax-project-select'].join(' ')

    unless opts.delete(:scope) == :all
      if @group
        opts['data-group-id'] = @group.id
      end
    end

    with_feature_enabled_data_attribute =
      case opts.delete(:with_feature_enabled)
      when 'issues'         then 'data-with-issues-enabled'
      when 'merge_requests' then 'data-with-merge-requests-enabled'
      end

    opts[with_feature_enabled_data_attribute] = true

    hidden_field_tag(id, opts[:selected], opts)
  end

  def select2_tag(id, opts = {})
    klass_opts = [opts[:class]]
    klass_opts << 'multiselect' if opts[:multiple]

    opts[:class] = klass_opts.join(' ')
    value = opts[:selected] || ''

    hidden_field_tag(id, value, opts)
  end
end

SelectsHelper.prepend_mod_with('SelectsHelper')
