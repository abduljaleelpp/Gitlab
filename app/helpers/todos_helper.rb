# frozen_string_literal: true

module TodosHelper
  def todos_pending_count
    @todos_pending_count ||= current_user.todos_pending_count
  end

  def todos_count_format(count)
    count > 99 ? '99+' : count.to_s
  end

  def todos_done_count
    @todos_done_count ||= current_user.todos_done_count
  end

  def todo_action_name(todo)
    case todo.action
    when Todo::ASSIGNED then todo.self_added? ? _('assigned') : _('assigned you')
    when Todo::REVIEW_REQUESTED then s_('Todos|requested a review of')
    when Todo::MENTIONED, Todo::DIRECTLY_ADDRESSED then format(
      s_("Todos|mentioned %{who} on"), who: todo_action_subject(todo)
    )
    when Todo::BUILD_FAILED then s_('Todos|The pipeline failed in')
    when Todo::MARKED then s_('Todos|added a todo for')
    when Todo::APPROVAL_REQUIRED then format(
      s_("Todos|set %{who} as an approver for"), who: todo_action_subject(todo)
    )
    when Todo::UNMERGEABLE then s_('Todos|Could not merge')
    when Todo::MERGE_TRAIN_REMOVED then s_("Todos|Removed from Merge Train:")
    end
  end

  def todo_self_addressing(todo)
    case todo.action
    when Todo::ASSIGNED then _('to yourself')
    when Todo::REVIEW_REQUESTED then _('from yourself')
    end
  end

  def todo_target_link(todo)
    text = raw(todo_target_type_name(todo) + ' ') +
      if todo.for_commit?
        content_tag(:span, todo.target_reference, class: 'commit-sha')
      else
        todo.target_reference
      end

    link_to text, todo_target_path(todo)
  end

  def todo_target_title(todo)
    # Design To Dos' filenames are displayed in `#todo_target_link` (see `Design#to_reference`),
    # so to avoid displaying duplicate filenames in the To Do list for designs,
    # we return an empty string here.
    return "" if todo.target.blank? || todo.for_design?

    "\"#{todo.target.title}\""
  end

  def todo_parent_path(todo)
    if todo.resource_parent.is_a?(Group)
      link_to todo.resource_parent.name, group_path(todo.resource_parent)
    else
      link_to_project(todo.project)
    end
  end

  def todo_target_type_name(todo)
    return _('design') if todo.for_design?
    return _('alert') if todo.for_alert?

    target_type = if todo.for_issue_or_work_item?
                    IntegrationsHelper.integration_issue_type(todo.target.issue_type)
                  else
                    IntegrationsHelper.integration_todo_target_type(todo.target_type)
                  end

    target_type.titleize.downcase
  end

  def todo_target_path(todo)
    return unless todo.target.present?

    path_options = todo_target_path_options(todo)

    if todo.for_commit?
      project_commit_path(todo.project, todo.target, path_options)
    elsif todo.for_design?
      todos_design_path(todo, path_options)
    elsif todo.for_alert?
      details_project_alert_management_path(todo.project, todo.target)
    elsif todo.for_issue_or_work_item?
      path_options[:only_path] = true
      Gitlab::UrlBuilder.build(todo.target, **path_options)
    else
      path = [todo.resource_parent, todo.target]

      path.unshift(:pipelines) if todo.build_failed?

      polymorphic_path(path, path_options)
    end
  end

  def todo_target_path_options(todo)
    { anchor: todo_target_path_anchor(todo) }
  end

  def todo_target_path_anchor(todo)
    dom_id(todo.note) if todo.note.present?
  end

  def todo_target_state_pill(todo)
    return unless show_todo_state?(todo)

    state = todo.target.state.to_s
    raw_state_to_i18n = {
      "closed" => _('Closed'),
      "merged" => _('Merged'),
      "resolved" => _('Resolved')
    }

    case todo.target
    when MergeRequest
      case state
      when 'closed'
        background_class = 'gl-bg-red-500'
      when 'merged'
        background_class = 'gl-bg-blue-500'
      end
    when Issue
      background_class = 'gl-bg-blue-500' if state == 'closed'
    when AlertManagement::Alert
      background_class = 'gl-bg-blue-500' if state == 'resolved'
    end

    tag.span class: "gl-my-0 gl-px-2 status-box #{background_class}" do
      raw_state_to_i18n[state] || state.capitalize
    end
  end

  def todos_filter_params
    {
      state: params[:state].presence,
      project_id: params[:project_id],
      author_id: params[:author_id],
      type: params[:type],
      action_id: params[:action_id]
    }.compact
  end

  def todos_filter_empty?
    todos_filter_params.values.none?
  end

  def no_todos_messages
    [
      s_('Todos|Good job! Looks like you don\'t have anything left on your To-Do List'),
      s_('Todos|Isn\'t an empty To-Do List beautiful?'),
      s_('Todos|Give yourself a pat on the back!'),
      s_('Todos|Nothing left to do. High five!'),
      s_('Todos|Henceforth, you shall be known as "To-Do Destroyer"')
    ]
  end

  def todos_filter_path(options = {})
    without = options.delete(:without)

    options = todos_filter_params.merge(options)

    if without.present?
      without.each do |key|
        options.delete(key)
      end
    end

    "#{request.path}?#{options.to_param}"
  end

  def todo_actions_options
    [
      { id: '', text: s_('Todos|Any Action') },
      { id: Todo::ASSIGNED, text: s_('Todos|Assigned') },
      { id: Todo::REVIEW_REQUESTED, text: s_('Todos|Review requested') },
      { id: Todo::MENTIONED, text: s_('Todos|Mentioned') },
      { id: Todo::MARKED, text: s_('Todos|Added') },
      { id: Todo::BUILD_FAILED, text: s_('Todos|Pipelines') }
    ]
  end

  def todo_types_options
    [
      { id: '', text: s_('Todos|Any Type') },
      { id: 'Issue', text: s_('Todos|Issue') },
      { id: 'MergeRequest', text: s_('Todos|Merge request') },
      { id: 'DesignManagement::Design', text: s_('Todos|Design') },
      { id: 'AlertManagement::Alert', text: s_('Todos|Alert') }
    ]
  end

  def todo_actions_dropdown_label(selected_action_id, default_action)
    selected_action = todo_actions_options.find { |action| action[:id] == selected_action_id.to_i }
    selected_action ? selected_action[:text] : default_action
  end

  def todo_types_dropdown_label(selected_type, default_type)
    selected_type = todo_types_options.find { |type| type[:id] == selected_type && type[:id] != '' }
    selected_type ? selected_type[:text] : default_type
  end

  def todo_due_date(todo)
    return unless todo.target.try(:due_date)

    is_due_today = todo.target.due_date.today?
    is_overdue = todo.target.overdue?
    css_class =
      if is_due_today
        'text-warning'
      elsif is_overdue
        'text-danger'
      else
        ''
      end

    content = content_tag(:span, class: css_class) do
      "Due #{is_due_today ? "today" : todo.target.due_date.to_s(:medium)}"
    end

    "&middot; #{content}".html_safe
  end

  def todo_author_display?(todo)
    !todo.build_failed? && !todo.unmergeable?
  end

  private

  def todos_design_path(todo, path_options)
    design = todo.target

    designs_project_issue_path(
      todo.resource_parent,
      design.issue,
      path_options.merge(
        vueroute: design.filename
      )
    )
  end

  def todo_action_subject(todo)
    todo.self_added? ? s_('Todos|yourself') : _('you')
  end

  def show_todo_state?(todo)
    case todo.target
    when MergeRequest, Issue
      %w(closed merged).include?(todo.target.state)
    when AlertManagement::Alert
      %i(resolved).include?(todo.target.state)
    else
      false
    end
  end
end

TodosHelper.prepend_mod_with('TodosHelper')
