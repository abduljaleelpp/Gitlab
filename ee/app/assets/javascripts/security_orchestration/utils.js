import createGqClient from '~/lib/graphql';
import { POLICY_TYPE_COMPONENT_OPTIONS } from './components/constants';

/**
 * Get a policy's type
 * @param {String} typeName policy's YAML manifest
 * @returns {String|null} policy type if available
 */
export const getPolicyType = (typeName = '') => {
  return Object.values(POLICY_TYPE_COMPONENT_OPTIONS).find(
    (component) => component.typeName === typeName,
  )?.value;
};

/**
 * Removes inital line dashes from a policy YAML that is received from the API, which
 * is not required for the user.
 * @param {String} manifest the policy from the API request
 * @returns {String} the policy without the initial dashes or the initial string
 */
export const removeUnnecessaryDashes = (manifest) => {
  return manifest.replace('---\n', '');
};

/**
 * Create GraphQL Client for threat monitoring
 */
export const gqClient = createGqClient();
