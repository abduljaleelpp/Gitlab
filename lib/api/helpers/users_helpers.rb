# frozen_string_literal: true

module API
  module Helpers
    module UsersHelpers
      extend ActiveSupport::Concern
      extend Grape::API::Helpers

      params :optional_params_ee do
      end

      params :optional_index_params_ee do
      end

      def model_error_messages(model)
        super.tap do |error_messages|
          # Remapping errors from nested associations.
          error_messages[:bio] = error_messages.delete(:"user_detail.bio") if error_messages.has_key?(:"user_detail.bio")
        end
      end

      # rubocop: disable CodeReuse/ActiveRecord
      def find_user_by_id(params)
        id = params[:user_id] || params[:id]
        User.find_by(id: id) || not_found!('User')
      end
      # rubocop: enable CodeReuse/ActiveRecord
    end
  end
end

API::Helpers::UsersHelpers.prepend_mod_with('API::Helpers::UsersHelpers')
