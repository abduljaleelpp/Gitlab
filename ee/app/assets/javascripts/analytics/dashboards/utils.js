import { isNumeric } from '~/lib/utils/number_utils';
import { fetchMetricsData } from '~/analytics/shared/utils';
import { METRICS_REQUESTS } from '~/cycle_analytics/constants';
import { DORA_METRICS } from './constants';

export const percentChange = ({ current, previous }) =>
  previous > 0 && current > 0 ? (current - previous) / previous : 0;

/**
 * Takes a flat array of metrics and extracts only the DORA metrics,
 * returning and key value object of the resulting metrics
 *
 * Currently DORA metrics are spread across 2 API endpoints, we fetch the data from
 * both then flatten it into a single array. This function takes that single array and
 * extracts the DORA specific metrics into a key value object.
 *
 * @param {Array} metrics - array of all the time / summary metrics
 * @returns {Object} an object with each of the 4 DORA metrics as a key and their relevant data
 */
export const extractDoraMetrics = (metrics = []) =>
  metrics
    .filter(({ identifier }) => Object.keys(DORA_METRICS).includes(identifier))
    .reduce((acc, curr) => {
      return {
        ...acc,
        [curr.identifier]: curr,
      };
    }, {});

/**
 * Fetches and merges DORA metrics into the given timePeriod objects.
 *
 * @param {Array} timePeriods - array of objects containing DORA metric values
 * @param {String} requestPath - URL path to use for the DORA metric API requests
 * @returns {Array} The original timePeriods array, with DORA metrics included
 */
export const fetchDoraMetrics = async ({ timePeriods, requestPath }) => {
  const promises = timePeriods.map((period) =>
    fetchMetricsData(METRICS_REQUESTS, requestPath, {
      created_after: period.start.toISOString(),
      created_before: period.end.toISOString(),
    }),
  );

  const results = await Promise.all(promises);
  return timePeriods.map((period, index) => ({
    ...period,
    ...extractDoraMetrics(results[index]),
  }));
};

/**
 * Takes an array of timePeriod objects containing DORA metrics, and returns
 * true if any of the timePeriods contain metric values > 0.
 *
 * @param {Array} timePeriods - array of objects containing DORA metric values
 * @returns {Boolean} true if there is any metric data, otherwise false.
 */
export const hasDoraMetricValues = (timePeriods) =>
  timePeriods.some((timePeriod) => {
    // timePeriod may contain more attributes than just the DORA metrics,
    // so filter out non-metrics before making a list of the raw values
    const metricValues = Object.entries(timePeriod)
      .filter(([k]) => Object.keys(DORA_METRICS).includes(k))
      .map(([, v]) => v.value);

    return metricValues.some((value) => isNumeric(value) && Number(value) > 0);
  });

/**
 * Takes N time periods of DORA metrics and generates the data rows
 * for the comparison table.
 *
 * @param {Array} timePeriods - Array of the DORA metrics for different time periods
 * @returns {Array} array comparing each DORA metric between the different time periods
 */
export const generateDoraTimePeriodComparisonTable = (timePeriods) => {
  return Object.entries(DORA_METRICS).map(([identifier, { label, formatValue }]) => {
    const data = { metric: { value: label } };
    timePeriods.forEach((timePeriod, index) => {
      // The last timePeriod is not rendered, we just use it
      // to determine the % change for the 2nd last timePeriod
      if (index === timePeriods.length - 1) return;

      const current = timePeriod[identifier];
      const previous = timePeriods[index + 1][identifier];

      data[timePeriod.key] = {
        value: current ? formatValue(current.value) : '-',
        change: percentChange({
          current: current?.value || 0,
          previous: previous?.value || 0,
        }),
      };
    });
    return data;
  });
};
