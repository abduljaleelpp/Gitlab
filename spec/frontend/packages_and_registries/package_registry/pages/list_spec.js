import { GlAlert, GlBanner, GlEmptyState, GlSprintf, GlLink } from '@gitlab/ui';
import Vue, { nextTick } from 'vue';

import VueApollo from 'vue-apollo';
import * as utils from '~/lib/utils/common_utils';
import { shallowMountExtended } from 'helpers/vue_test_utils_helper';
import createMockApollo from 'helpers/mock_apollo_helper';
import waitForPromises from 'helpers/wait_for_promises';
import { stubComponent } from 'helpers/stub_component';
import ListPage from '~/packages_and_registries/package_registry/pages/list.vue';
import PackageTitle from '~/packages_and_registries/package_registry/components/list/package_title.vue';
import PackageSearch from '~/packages_and_registries/package_registry/components/list/package_search.vue';
import OriginalPackageList from '~/packages_and_registries/package_registry/components/list/packages_list.vue';
import DeletePackage from '~/packages_and_registries/package_registry/components/functional/delete_package.vue';
import DeleteModal from '~/packages_and_registries/package_registry/components/delete_modal.vue';
import {
  PROJECT_RESOURCE_TYPE,
  GROUP_RESOURCE_TYPE,
  GRAPHQL_PAGE_SIZE,
  HIDE_PACKAGE_MIGRATION_SURVEY_COOKIE,
  EMPTY_LIST_HELP_URL,
  PACKAGE_HELP_URL,
  DELETE_PACKAGES_ERROR_MESSAGE,
  DELETE_PACKAGES_SUCCESS_MESSAGE,
} from '~/packages_and_registries/package_registry/constants';

import getPackagesQuery from '~/packages_and_registries/package_registry/graphql/queries/get_packages.query.graphql';
import destroyPackagesMutation from '~/packages_and_registries/package_registry/graphql/mutations/destroy_packages.mutation.graphql';
import {
  packagesListQuery,
  packageData,
  pagination,
  packagesDestroyMutation,
  packagesDestroyMutationError,
} from '../mock_data';

jest.mock('~/flash');

describe('PackagesListApp', () => {
  let wrapper;
  let apolloProvider;

  const defaultProvide = {
    emptyListIllustration: 'emptyListIllustration',
    isGroupPage: true,
    fullPath: 'gitlab-org',
  };

  const PackageList = {
    name: 'package-list',
    template: '<div><slot name="empty-state"></slot></div>',
    props: OriginalPackageList.props,
  };
  const GlLoadingIcon = { name: 'gl-loading-icon', template: '<div>loading</div>' };

  const searchPayload = {
    sort: 'VERSION_DESC',
    filters: { packageName: 'foo', packageType: 'CONAN' },
  };

  const findAlert = () => wrapper.findComponent(GlAlert);
  const findBanner = () => wrapper.findComponent(GlBanner);
  const findPackageTitle = () => wrapper.findComponent(PackageTitle);
  const findSearch = () => wrapper.findComponent(PackageSearch);
  const findListComponent = () => wrapper.findComponent(PackageList);
  const findEmptyState = () => wrapper.findComponent(GlEmptyState);
  const findDeletePackage = () => wrapper.findComponent(DeletePackage);
  const findDeletePackagesModal = () => wrapper.findComponent(DeleteModal);

  const mountComponent = ({
    resolver = jest.fn().mockResolvedValue(packagesListQuery()),
    mutationResolver,
    provide = defaultProvide,
  } = {}) => {
    Vue.use(VueApollo);

    const requestHandlers = [
      [getPackagesQuery, resolver],
      [destroyPackagesMutation, mutationResolver],
    ];
    apolloProvider = createMockApollo(requestHandlers);

    wrapper = shallowMountExtended(ListPage, {
      apolloProvider,
      provide,
      stubs: {
        GlBanner,
        GlEmptyState,
        GlLoadingIcon,
        GlSprintf,
        GlLink,
        PackageList,
        DeletePackage,
        DeleteModal: stubComponent(DeleteModal, {
          methods: {
            show: jest.fn(),
          },
        }),
      },
    });
  };

  afterEach(() => {
    wrapper.destroy();
  });

  const waitForFirstRequest = async () => {
    // emit a search update so the query is executed
    findSearch().vm.$emit('update', { sort: 'NAME_DESC', filters: [] });
    return waitForPromises();
  };

  it('does not execute the query without sort being set', () => {
    const resolver = jest.fn().mockResolvedValue(packagesListQuery());

    mountComponent({ resolver });

    expect(resolver).not.toHaveBeenCalled();
  });

  it('renders', async () => {
    mountComponent();

    await waitForFirstRequest();

    expect(wrapper.element).toMatchSnapshot();
  });

  it('has a package title', async () => {
    mountComponent();

    await waitForFirstRequest();

    expect(findPackageTitle().exists()).toBe(true);
    expect(findPackageTitle().props()).toMatchObject({
      count: 2,
      helpUrl: PACKAGE_HELP_URL,
    });
  });

  describe('package migration survey banner', () => {
    describe('with no cookie set', () => {
      beforeEach(() => {
        utils.setCookie = jest.fn();

        mountComponent();
      });

      it('displays the banner', () => {
        expect(findBanner().exists()).toBe(true);
      });

      it('does not call setCookie', () => {
        expect(utils.setCookie).not.toHaveBeenCalled();
      });

      describe('when the close button is clicked', () => {
        beforeEach(() => {
          findBanner().vm.$emit('close');
        });

        it('sets the dismissed cookie', () => {
          expect(utils.setCookie).toHaveBeenCalledWith(
            HIDE_PACKAGE_MIGRATION_SURVEY_COOKIE,
            'true',
          );
        });

        it('does not display the banner', () => {
          expect(findBanner().exists()).toBe(false);
        });
      });

      describe('when the primary button is clicked', () => {
        beforeEach(() => {
          findBanner().vm.$emit('primary');
        });

        it('sets the dismissed cookie', () => {
          expect(utils.setCookie).toHaveBeenCalledWith(
            HIDE_PACKAGE_MIGRATION_SURVEY_COOKIE,
            'true',
          );
        });

        it('does not display the banner', () => {
          expect(findBanner().exists()).toBe(false);
        });
      });
    });

    describe('with the dismissed cookie set', () => {
      beforeEach(() => {
        jest.spyOn(utils, 'getCookie').mockReturnValue('true');

        mountComponent();
      });

      it('does not display the banner', () => {
        expect(findBanner().exists()).toBe(false);
      });
    });
  });

  describe('search component', () => {
    it('exists', () => {
      mountComponent();

      expect(findSearch().exists()).toBe(true);
    });

    it('on update triggers a new query with updated values', async () => {
      const resolver = jest.fn().mockResolvedValue(packagesListQuery());
      mountComponent({ resolver });

      findSearch().vm.$emit('update', searchPayload);

      await waitForPromises();

      expect(resolver).toHaveBeenCalledWith(
        expect.objectContaining({
          groupSort: searchPayload.sort,
          ...searchPayload.filters,
        }),
      );
    });
  });

  describe('list component', () => {
    let resolver;

    beforeEach(() => {
      resolver = jest.fn().mockResolvedValue(packagesListQuery());
      mountComponent({ resolver });
    });

    it('exists and has the right props', async () => {
      await waitForFirstRequest();
      expect(findListComponent().props()).toMatchObject({
        list: expect.arrayContaining([expect.objectContaining({ id: packageData().id })]),
        isLoading: false,
        pageInfo: expect.objectContaining({ endCursor: pagination().endCursor }),
      });
    });

    it('when list emits next-page fetches the next set of records', async () => {
      await waitForFirstRequest();
      findListComponent().vm.$emit('next-page');
      await waitForPromises();

      expect(resolver).toHaveBeenCalledWith(
        expect.objectContaining({ after: pagination().endCursor, first: GRAPHQL_PAGE_SIZE }),
      );
    });

    it('when list emits prev-page fetches the prev set of records', async () => {
      await waitForFirstRequest();
      findListComponent().vm.$emit('prev-page');
      await waitForPromises();

      expect(resolver).toHaveBeenCalledWith(
        expect.objectContaining({ before: pagination().startCursor, last: GRAPHQL_PAGE_SIZE }),
      );
    });
  });

  describe.each`
    type                     | sortType
    ${PROJECT_RESOURCE_TYPE} | ${'sort'}
    ${GROUP_RESOURCE_TYPE}   | ${'groupSort'}
  `('$type query', ({ type, sortType }) => {
    let provide;
    let resolver;

    const isGroupPage = type === GROUP_RESOURCE_TYPE;

    beforeEach(() => {
      provide = { ...defaultProvide, isGroupPage };
      resolver = jest.fn().mockResolvedValue(packagesListQuery({ type }));
      mountComponent({ provide, resolver });
      return waitForFirstRequest();
    });

    it('succeeds', () => {
      expect(findPackageTitle().props('count')).toBe(2);
    });

    it('calls the resolver with the right parameters', () => {
      expect(resolver).toHaveBeenCalledWith(
        expect.objectContaining({ isGroupPage, [sortType]: 'NAME_DESC' }),
      );
    });
  });

  describe('empty state', () => {
    beforeEach(() => {
      const resolver = jest.fn().mockResolvedValue(packagesListQuery({ extend: { nodes: [] } }));
      mountComponent({ resolver });

      return waitForFirstRequest();
    });
    it('generate the correct empty list link', () => {
      const link = findListComponent().findComponent(GlLink);

      expect(link.attributes('href')).toBe(EMPTY_LIST_HELP_URL);
      expect(link.text()).toBe('publish and share your packages');
    });

    it('includes the right content on the default tab', () => {
      expect(findEmptyState().text()).toContain(ListPage.i18n.emptyPageTitle);
    });
  });

  describe('filter without results', () => {
    beforeEach(async () => {
      mountComponent();

      await waitForFirstRequest();

      findSearch().vm.$emit('update', searchPayload);

      return nextTick();
    });

    it('should show specific empty message', () => {
      expect(findEmptyState().text()).toContain(ListPage.i18n.noResultsTitle);
      expect(findEmptyState().text()).toContain(ListPage.i18n.widenFilters);
    });
  });

  describe('delete package', () => {
    it('exists and has the correct props', async () => {
      mountComponent();

      await waitForFirstRequest();

      expect(findDeletePackage().props()).toMatchObject({
        refetchQueries: [{ query: getPackagesQuery, variables: {} }],
        showSuccessAlert: true,
      });
    });

    it('deletePackage is bound to package-list package:delete event', async () => {
      mountComponent();

      await waitForFirstRequest();

      findListComponent().vm.$emit('package:delete', { id: 1 });

      expect(findDeletePackage().emitted('start')).toEqual([[]]);
    });

    it('start and end event set loading correctly', async () => {
      mountComponent();

      await waitForFirstRequest();

      findDeletePackage().vm.$emit('start');

      await nextTick();

      expect(findListComponent().props('isLoading')).toBe(true);

      findDeletePackage().vm.$emit('end');

      await nextTick();

      expect(findListComponent().props('isLoading')).toBe(false);
    });
  });

  describe('bulk delete package', () => {
    const items = [{ id: '1' }, { id: '2' }];

    it('deletePackage is bound to package-list package:delete event', async () => {
      mountComponent();

      await waitForFirstRequest();

      findListComponent().vm.$emit('delete', [{ id: '1' }, { id: '2' }]);

      await waitForPromises();

      expect(findDeletePackagesModal().props('itemsToBeDeleted')).toEqual(items);
    });

    it('calls mutation with the right values and shows success alert', async () => {
      const mutationResolver = jest.fn().mockResolvedValue(packagesDestroyMutation());
      mountComponent({
        mutationResolver,
      });

      await waitForFirstRequest();

      findListComponent().vm.$emit('delete', items);

      findDeletePackagesModal().vm.$emit('confirm');

      expect(mutationResolver).toHaveBeenCalledWith({
        ids: items.map((item) => item.id),
      });

      await waitForPromises();

      expect(findAlert().exists()).toBe(true);
      expect(findAlert().props('variant')).toEqual('success');
      expect(findAlert().text()).toMatchInterpolatedText(DELETE_PACKAGES_SUCCESS_MESSAGE);
    });

    it('on error shows danger alert', async () => {
      const mutationResolver = jest.fn().mockResolvedValue(packagesDestroyMutationError());
      mountComponent({
        mutationResolver,
      });

      await waitForFirstRequest();

      findListComponent().vm.$emit('delete', items);

      findDeletePackagesModal().vm.$emit('confirm');

      await waitForPromises();

      expect(findAlert().exists()).toBe(true);
      expect(findAlert().props('variant')).toEqual('danger');
      expect(findAlert().text()).toMatchInterpolatedText(DELETE_PACKAGES_ERROR_MESSAGE);
    });
  });
});
