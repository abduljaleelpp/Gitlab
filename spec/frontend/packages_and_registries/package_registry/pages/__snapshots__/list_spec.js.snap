// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PackagesListApp renders 1`] = `
<div>
  <!---->
   
  <gl-card-stub
    bodyclass="gl-display-flex gl-p-0!"
    class="gl-px-8 gl-py-6 gl-line-height-20 gl-mt-3"
    footerclass=""
    headerclass=""
  >
    <!---->
     
    <div
      class="gl-banner-content"
    >
      <h2
        class="gl-banner-title"
      >
        Help us learn about your registry migration needs
      </h2>
       
      <p>
        If you are interested in migrating packages from your private registry to the GitLab Package Registry, take our survey and tell us more about your needs.
      </p>
       
      <gl-button-stub
        buttontextclasses=""
        category="primary"
        data-testid="gl-banner-primary-button"
        href="https://gitlab.fra1.qualtrics.com/jfe/form/SV_cHomH9FPzOaiDTU"
        icon=""
        size="medium"
        variant="confirm"
      >
        Take survey
      </gl-button-stub>
       
    </div>
     
    <gl-button-stub
      aria-label="Close banner"
      buttontextclasses=""
      category="tertiary"
      class="gl-banner-close"
      icon="close"
      size="small"
      variant="default"
    />
  </gl-card-stub>
   
  <package-title-stub
    count="2"
    helpurl="/help/user/packages/index"
  />
   
  <package-search-stub
    class="gl-mb-5"
  />
   
  <div>
    <section
      class="gl-display-flex empty-state gl-text-center gl-flex-direction-column"
    >
      <div
        class="gl-max-w-full"
      >
        <div
          class="svg-250 svg-content"
        >
          <img
            alt=""
            class="gl-max-w-full gl-dark-invert-keep-hue"
            role="img"
            src="emptyListIllustration"
          />
        </div>
      </div>
       
      <div
        class="gl-max-w-full gl-m-auto"
      >
        <div
          class="gl-mx-auto gl-my-0 gl-p-5"
        >
          <h1
            class="gl-font-size-h-display gl-line-height-36 h4"
          >
            
          There are no packages yet
        
          </h1>
           
          <p
            class="gl-mt-3"
          >
            Learn how to 
            <b-link-stub
              class="gl-link"
              event="click"
              href="/help/user/packages/package_registry/index"
              routertag="a"
              target="_blank"
            >
              publish and share your packages
            </b-link-stub>
             with GitLab.
          </p>
           
          <div
            class="gl-display-flex gl-flex-wrap gl-justify-content-center"
          >
            <!---->
             
            <!---->
          </div>
        </div>
      </div>
    </section>
  </div>
   
  <div />
</div>
`;
